File: presentation/src/main/java/by/grodno/vasili/githubsimpleapp/feature/users/UserItemDatasource.java
package by.grodno.vasili.githubsimpleapp.feature.users;

import java.util.List;

import androidx.annotation.NonNull;
import androidx.paging.PageKeyedDataSource;
import by.grodno.vasili.domain.interactor.GetUsersUseCase;
import by.grodno.vasili.domain.model.User;
import io.reactivex.observers.DisposableSingleObserver;
import timber.log.Timber;

/**
 * Datasource for pagination
 */
public class UserItemDatasource extends PageKeyedDataSource<String, UserItem> {
    private static final int START_POSITION = 0;
    private final GetUsersUseCase getUsersUseCase;
    private final UserItemMapper mapper;

    UserItemDatasource(GetUsersUseCase getUsersUseCase, UserItemMapper mapper) {
        this.getUsersUseCase = getUsersUseCase;
        this.mapper = mapper;
    }

    @Override
    public void loadInitial(@NonNull LoadInitialParams<String> params, @NonNull LoadInitialCallback<String, UserItem> callback) {
        DisposableSingleObserver<List<User>> observer = new DisposableSingleObserver<List<User>>() {
            @Override
            public void onSuccess(List<User> users) {
                callback.onResult(mapper.mapList(users), null, lastUserId(users));
            }

            @Override
            public void onError(Throwable e) {
                Timber.e(e, "Error while retrieving Users");
            }
        };

        getUsersUseCase.execute(observer, GetUsersUseCase.Params.create(START_POSITION));
    }

    @Override
    public void loadBefore(@NonNull LoadParams<String> params, @NonNull LoadCallback<String, UserItem> callback) {
    }

    @Override
    public void loadAfter(@NonNull LoadParams<String> params, @NonNull LoadCallback<String, UserItem> callback) {
        DisposableSingleObserver<List<User>> observer = new DisposableSingleObserver<List<User>>() {
            @Override
            public void onSuccess(List<User> users) {
                callback.onResult(mapper.mapList(users), lastUserId(users));
            }

            @Override
            public void onError(Throwable e) {
                Timber.e(e, "Error while retrieving Users");
            }
        };
        getUsersUseCase.execute(observer, GetUsersUseCase.Params.create(Integer.parseInt(params.key)));
    }

    private String lastUserId(List<User> users) {
        return users.get(users.size() - 1).id;
    }
}

File: presentation/src/main/java/by/grodno/vasili/githubsimpleapp/feature/users/UserItemDatasource.kt
package by.grodno.vasili.githubsimpleapp.feature.users

import androidx.paging.PageKeyedDataSource
import by.grodno.vasili.domain.interactor.GetUsersUseCase
import by.grodno.vasili.domain.model.User
import io.reactivex.observers.DisposableSingleObserver
import timber.log.Timber

/**
 * Datasource for pagination.
 */
class UserItemDatasource(private val getUsersUseCase: GetUsersUseCase) : PageKeyedDataSource<String, UserItem>() {
    override fun loadInitial(params: LoadInitialParams<String>, callback: LoadInitialCallback<String, UserItem>) {
        val observer: DisposableSingleObserver<List<User>> = object : DisposableSingleObserver<List<User>>() {
            override fun onSuccess(users: List<User>) {
                callback.onResult(users.map { UserItem(it) }, null, lastUserId(users))
            }

            override fun onError(e: Throwable) {
                Timber.e(e, "Error while retrieving Users")
            }
        }
        getUsersUseCase.execute(observer, GetUsersUseCase.Params.create(START_POSITION))
    }

    override fun loadBefore(params: LoadParams<String>, callback: LoadCallback<String, UserItem>) {}
    override fun loadAfter(params: LoadParams<String>, callback: LoadCallback<String, UserItem>) {
        val observer: DisposableSingleObserver<List<User>> = object : DisposableSingleObserver<List<User>>() {
            override fun onSuccess(users: List<User>) {
                callback.onResult(users.map { UserItem(it) }, lastUserId(users))
            }

            override fun onError(e: Throwable) {
                Timber.e(e, "Error while retrieving Users")
            }
        }
        getUsersUseCase.execute(observer, GetUsersUseCase.Params.create(params.key.toInt()))
    }

    private fun lastUserId(users: List<User>): String {
        return users[users.size - 1].id
    }

    companion object {
        private const val START_POSITION = 0
    }
}
